### DISCUSSION TITLE
set-output deprecated

### DISCUSSION THREAD

**Discussion by funes79** [community/community#35994] [Actions] [answered]
Two days ago the Github Actions announced the deprecation of set-output and workflows started to emit these warnings:
```
The `set-output` command is deprecated and will be disabled soon. Please upgrade to using Environment Files. For more information see: https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/
```

However there is no clear documentation how to migrate the workflows and how to rewrite to use env vars - whether to use GITHUB_ENV or GITHUB_OUTPUT.

I believe `set-output` was used widely in many flows, so it deserves a more detailed documentation and migration guide.

**Comment by airtower-luna** [comment]
The documentation on how to set outputs now is here: https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-output-parameter You do **not** need to replace outputs with environment variables, just set the outputs by writing to a file instead of stdout.

One thing that's missing there are multiline outputs, I found by experimenting that you can set multiline outputs [using the same format as multiline environment variables](https://github.com/airtower-luna/hello-github-actions/commit/6c7cd2afa0c249b2c909c06e5ff62887ed4ab6b6).

**Reply by deitry** [reply]
I tried new approach and discovered it does not work on windows runners. Please see corresponding runs on Ubuntu:
https://github.com/deitry/test-output/actions/runs/3249439567/jobs/5331843416

and Windows:
https://github.com/deitry/test-output/actions/runs/3249439554/jobs/5331843290

```yml
    steps:
      - id: step1
        run: echo "test=Hello world" >> $GITHUB_OUTPUT

      - run: echo "${{ steps.step1.outputs.test }}"
```

**Reply by chrisnruud** [reply]
To use multiline strings I had to do something like this:

```bash
          COMMENT_WITH_MULTILINE_MARKDOWN="${COMMENT_TITLE}
          #### Section 1
          #### Section 2
          #### Section 3
          #### Section 4
          \`\`\`bash
          ${VARIABLE_HOLDING_RESULT_OF_A_CAT_FROM_A_SCRIPT_FILE_ALSO_MULTILINE}
          \`\`\`"
          echo 'comment<<RANDOM_DELIMITER' >> $GITHUB_OUTPUT
          echo "${COMMENT_WITH_MULTILINE_MARKDOWN}" >> $GITHUB_OUTPUT
          echo 'RANDOM_DELIMITER' >> $GITHUB_OUTPUT
```

**Reply by codeaphex** [reply]
> I tried new approach and discovered it does not work on windows runners. Please see corresponding runs on Ubuntu: https://github.com/deitry/test-output/actions/runs/3249439567/jobs/5331843416
> 
> and Windows: https://github.com/deitry/test-output/actions/runs/3249439554/jobs/5331843290
> 
> ```yaml
>     steps:
>       - id: step1
>         run: echo "test=Hello world" >> $GITHUB_OUTPUT
> 
>       - run: echo "${{ steps.step1.outputs.test }}"
> ```

Hey just ran into the same problem until I found the powershell switch on top of the docs:
![image](https://user-images.githubusercontent.com/39407721/196041204-2c3c0d81-2832-434b-a1fa-f32671c05c08.png)
Correct usage on windows/powershell would be:
```yaml
run: echo "test=Hello world" >> $env:GITHUB_OUTPUT
```

**Reply by deitry** [reply]
OMG, thanks, I totally missed that

**Reply by Justin-JHG** [reply]
same here, thank you so much

**Reply by ptfw** [reply]
@codeaphex  where are you finding that doc ? I was working off these docs that don't have a PowerShell tab  https://docs.github.com/en/actions/using-jobs/defining-outputs-for-jobs

**Reply by codeaphex** [reply]
https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions?tool=powershell#setting-an-output-parameter

**Comment by rgoodwin** [comment]
Thank you for the tip. The docs are pretty vague on how this works.

**Comment by kcgen** [comment, marked as answer]
## Automatic conversion one-liner

Changelog:
 - 2022-Oct-17: Updated to handle @funes79's complex AWS line below. Thanks for flagging this @funes79!

If you're using `echo "::set-output name=key::val"` syntax, this GNU sed one liner should get you most of the way: 

#### Bash-based workflows:
 
``` shell
sed \
  -i \
  '/set-output/ { 
    s/::set-output name=//;
    s/::/=/;
    s/"$/" >> $GITHUB_OUTPUT/;
  }' \
  .github/workflows/bash-workflows.yml
```

#### Powershell-based workflows:
 
``` shell
sed \
  -i \
  '/set-output/ { 
    s/::set-output name=//;
    s/::/=/;
    s/"$/" >> $env:GITHUB_OUTPUT/;
  }' \
  .github/workflows/powershell-workflows.yml
```


After running it, inspect the `git diff` and make any final touches that aren't perfect.

---

Here's an example of what it will convert:

```diff
         shell: bash
         run: |
           mkdir -p "${CCACHE_DIR}"
-          echo "::set-output name=dir::$CCACHE_DIR"
-          echo "::set-output name=today::$(date -I)"
-          echo "::set-output name=yesterday::$(date --date=yesterday -I)"
-          echo "::set-output name=name_hash::$(echo '${{ matrix.conf.name }}' | shasum | cut -b-8)"
+          echo "dir=$CCACHE_DIR" >> $GITHUB_OUTPUT
+          echo "today=$(date -I)" >> $GITHUB_OUTPUT
+          echo "yesterday=$(date --date=yesterday -I)" >> $GITHUB_OUTPUT
+          echo "name_hash=$(echo '${{ matrix.conf.name }}' | shasum | cut -b-8)" >> $GITHUB_OUTPUT
```

**Reply by deitry** [reply]
For my knowing new way seems more correct, since in PowerShell you have to use call operator `&` to run `.exe` by path.
https://stackoverflow.com/questions/4639894/executing-an-exe-file-using-a-powershell-script

**Reply by funes79** [reply]
The automatic conversion does not work if the code has expression
`echo "::set-output name=key::$(expression)"`
and also in case `##` is used:
`echo "##[set-output name=branch;]$(echo ${GITHUB_REF#refs/heads/})"`

**Reply by kcgen** [reply]
@funes79 ,

> The automatic conversion does not work if the code has expression
echo "::set-output name=key::$(expression)"

It handles that exact expression; can you paste your specific instance?

![2022-10-17_07-22](https://user-images.githubusercontent.com/1557255/196202332-8ae51da7-cb0b-4e5e-bb92-d47d1f70c269.png)

> and also in case ## is used:
echo "##[set-output name=branch;]$(echo ${GITHUB_REF#refs/heads/})"

It also handles that exact expression:

![2022-10-17_07-25](https://user-images.githubusercontent.com/1557255/196203165-f6cb6375-9ca1-424a-8be2-5bdc889c5c05.png)

**Reply by funes79** [reply]
```
echo "::set-output name=aws_role::$(if [[ "$B" == "main" ]]; then echo 'arn:aws:iam::12345679012:role/github_actions_admin'; else echo 'arn:aws:iam::999999999:role/github_ssm_role'; fi )"
```
translates to:
```
echo "aws_role=$(if [[ "$B" == "main" ]]; then echo 'arn:aws:iam=12345679012:role/github_actions_admin'; else echo 'arn:aws:iam=999999999:role/github_ssm_role'; fi )"      
```
Please note that the double `::` in AWS arn was replaced by `=` 

The timezone:
![image](https://user-images.githubusercontent.com/20202121/196212235-b9c31e86-9757-4d64-ae2b-f35425d33f36.png)

**Reply by kcgen** [reply]
Thanks for posting your case, @funes79. That's quite an example! 
I've updated the original post and mentioned the change.
When you have a chance to try the update, it would be good to know that it now handles it.

**Comment by sayhiben** [comment]
For what it's worth, this change materially impacts where you can set output from. Perhaps that's a good thing or design intention. Regardless, any scripts or outputs that previously ran in a Docker container will now need to volume-mount the $GITHUB_OUTPUT file (and likely will want to set it as an env var), whereas previous implementations allowed stdout to be intercepted at the top level

**Comment by Abbasisa** [comment]
> Two days ago the Github Actions announced the deprecation of set-output and workflows started to emit these warnings:
> 
> ```
> The `set-output` command is deprecated and will be disabled soon. Please upgrade to using Environment Files. For more information see: https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/
> ```
> 
> However there is no clear documentation how to migrate the workflows and how to rewrite to use env vars - whether to use GITHUB_ENV or GITHUB_OUTPUT.
> 
> I believe `set-output` was used widely in many flows, so it deserves a more detailed documentation and migration guide.
35994

**Comment by Abbasisa** [comment]
#35994
![Screenshot_20221017_205658](https://user-images.githubusercontent.com/116004766/196229615-a5e80e4c-afeb-48ff-8d46-8eeb5c36432f.jpg)

**Comment by Bykerone** [comment]
Enviado desde mi HUAWEI P30 lite-------- Mensaje original --------De: Justin Liu ***@***.***>Fecha: mié., 19 de octubre de 2022 7:33 p. m.Para: community/community ***@***.***>CC: Subscribed ***@***.***>Asunto: Re: [community/community] set-output deprecated (Discussion #35994)
same here, thank you so much

**Comment by funes79** [comment]
The port of yaml files by replacing the set-output to an echo redirect to GITHUB_OUTPUT works. But my python code does not log correclty the multiline values:


```
def encode_chars(value):
    return value.replace('%', '%25').replace('\n', '%0A').replace('\r', '%0D')

def set_output(name, value):
    if isinstance(value, str):
        value = encode_chars(value)
    if 'GITHUB_OUTPUT' in os.environ:
        with open(os.environ['GITHUB_OUTPUT'], "a") as ff:
            ff.write(f'{name}<<EEOOFF\n')
            ff.write(f'{value}')
            ff.write('\nEEOOFF\n')
    else:
        print(f'::set-output name={name}::{value}')
```

Resulting in a non replaced new line characters in Issues :
![image](https://user-images.githubusercontent.com/20202121/198070091-e74a3004-d7fa-41ff-bdd0-38901be2efae.png)

So it looks like that the special encoding is not needed:
```
def set_output(name, value):
    if 'GITHUB_OUTPUT' in os.environ:
        with open(os.environ['GITHUB_OUTPUT'], "a") as ff:
            ff.write(f'{name}<<EEOOFF\n')
            ff.write(f'{value}')
            ff.write('\nEEOOFF\n')
    else:
        if isinstance(value, str):
            value = encode_chars(value)
        print(f'::set-output name={name}::{value}')
```

**Reply by airtower-luna** [reply]
Exactly. :slightly_smiling_face: 

Also it's [strongly recommended](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#multiline-strings) (red warning box) to randomly generate the delimiter, to prevent anyone from injecting a predictable one (`EOF` and `EEOOFF` are equally predictable). [I simply use `secrets.token_hex()`](https://github.com/airtower-luna/hello-github-actions/blob/59f2f7386549f6efeb7e2d1d01ec2cd896983c2f/.github/workflows/multiline-output.yaml#L21) for that. How critical it is depends on what kind of data goes into your outputs.

**Comment by ma-caylent** [comment]
same issue as @funes79 faced regarding non-replaced new line characters. any fix for this? :pray:

**Reply by airtower-luna** [reply]
You need to use the same format as for [multiline environment variables](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#multiline-strings), obviously writing to the outputs file instead. Mind that you should generate a random delimiter during each run.

**Comment by bstordrup** [comment]
All examples on how to use the recommended show code in either yaml or in PowerShell.

What about GitHub actions written in .NET? The previous attempt (as demonstrated by the [GitHub Action in .NET sample](https://github.com/dotnet/samples/blob/main/github-actions/DotNet.GitHubAction/DotNet.GitHubAction/Program.cs)) made use of a Console.WriteLine like this:

```
    Console.WriteLine($"::set-output name=updated-metrics::{updatedMetrics}");
    Console.WriteLine($"::set-output name=summary-title::{title}");
    Console.WriteLine($"::set-output name=summary-details::{summary}");
```

What would the correct way be to do the same with $env.GITHUB_OUTPUT

**Comment by bstordrup** [comment]
Figured it out. The way to do it is:

```
    var githubOutputFile = Environment.GetEnvironmentVariable("GITHUB_OUTPUT", EnvironmentVariableTarget.Process);
    if (!string.IsNullOrWhiteSpace(githubOutputFile))
    {
        using (var textWriter = new StreamWriter(githubOutputFile!, true, Encoding.UTF8))
        {
            textWriter.WriteLine($"updated-metrics={updatedMetrics}");
            textWriter.WriteLine($"summary-title={title}");
            textWriter.WriteLine($"summary-details={summary}");
        }
    }
    else
    {
        Console.WriteLine($"::set-output name=updated-metrics::{updatedMetrics}");
        Console.WriteLine($"::set-output name=summary-title::{title}");
        Console.WriteLine($"::set-output name=summary-details::{summary}");
    }
```

**Comment by johnrhunt** [comment]
Why doesn't github just intelligently translate this on the fly?

**Comment by M97Chahboun** [comment]
[Here Alternative way for use terminal or any programming language output on workflow
](https://medium.com/@m97chahboun/how-to-use-any-programming-language-on-github-workflow-f5e389cfd886)
